### PR TITLE
Feature/snapshot batch update

### PR DESCRIPTION
**Pull request by dabasov**
## Changes
Snapshot logic implemented

## Associated PRs (Link as appropriate):
- blobber:
- gosdk: feature/snapshot
- system_test:
- zboxcli:
- zwalletcli:
- 0box: feature/snapshot

### CODE REVIEW

File: code/go/0chain.net/chaincore/chain/state.go [hunk 1]
```diff
@@ -773,3 +781,30 @@ func (c *Chain) emitUserEvent(sc bcstate.StateContextI, usr *event.User) error {
...
 	return nil
 }
+func (c *Chain) emitMintEvent(sc bcstate.StateContextI, usr *event.User) error {
+	if c.GetEventDb() == nil {
+		return nil
+	}
+
+	sc.EmitEvent(event.TypeStats, event.TagAddMint, usr.UserID, usr)
+
+	return nil
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/chain/state.go]: We can remove the return error as there's no error actually, callers should also be updated to remove the error checking code. Same as the other event emit methods below.

File: code/go/0chain.net/chaincore/httpclientutil/http_client_util.go [hunk 2]
```diff
@@ -755,7 +755,7 @@ func SendSmartContractTxn(txn *Transaction,
 	txn.ToClientID = address
 	txn.Value = value
 	txn.Fee = fee
-	txn.TransactionType = TxnTypeSmartContract
+	txn.TransactionType = TxnTypeStats
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/httpclientutil/http_client_util.go]: TxnTypeSmartContract is better I think, it is a smart contract, not a stats.

File: code/go/0chain.net/chaincore/transaction/type.go [hunk 3]
```diff
@@ -13,7 +13,7 @@ const (
 
 	TxnTypeData = 10 // A transaction to just store a piece of data on the block chain
 
-	TxnTypeSmartContract = 1000 // A smart contract transaction type
+	TxnTypeStats = 1000 // A smart contract transaction type
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/transaction/type.go]: Please use TxnTypeSmartContract

File: code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+}
+
+func paginate(round int64, pageAmount int64, count int64) (int64, int64, int) {
+	size := int64(math.Ceil(float64(count) / float64(pageAmount)))
+	currentPageNumber := round % pageAmount
+
+	currentPageSize := size
+	if pageAmount-currentPageNumber == 1 { //last page
+		currentPageSize = count - size*currentPageNumber
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go]: The currentPageSize could be greater than the size, is that the correct way to calculate the currentPageSize? Suppose the count is 11, pageAmount is 2, round = 101, then size will be 3, currentPageNumber = 1, and currentPageSize = 11 - 3*1 = 8, which is greater than the size 3.

File: code/go/0chain.net/smartcontract/dbs/event/snapshot.go [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+//unallocated - this is equal to (staked - allocated) and allows clients to purchase allocations with free space blobbers.
+//used - this is the actual usage or data that is in the server.
+//staked + unstaked = max_capacity
+//allocated + unallocated = staked
+
+// swagger:model Snapshot
+type Snapshot struct {
+	Round int64 `gorm:"primaryKey;autoIncrement:false" json:"round"`
+
+	TotalMint            int64 `json:"total_mint"`
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/snapshot.go:25]: Please use currency.Coin for tokens, and use currency package functions for tokens calculation. Same as other fields below

File: docker.local/config/0chain.yaml [hunk 6]
```diff
@@ -3,7 +3,7 @@ version: 1.0
 logging:
   level: "debug"
   verbose: false
-  console: false # printing log to console is only supported in development mode
+  console: true # printing log to console is only supported in development mode
```
Comment (peterlimg) [inline docker.local/config/0chain.yaml]: Let's turn off console logs for production

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 7]
```diff
@@ -23,9 +23,14 @@ type (
...
 	TypeNone EventType = iota
 	TypeError
+	TypeChain
 	TypeStats
 )
 
+const GB = 1024 * 1024 * 1024
+const period = 10
+const pageLimit = int64(50)
+
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/dbs/event/process.go:33]: In https://github.com/0chain/0chain/pull/1777 I am creating a  [settings object](https://github.com/0chain/0chain/blob/88f91752f55d8957200fc07dab05cada98aa8b98/code/go/0chain.net/smartcontract/dbs/event/settings.go) as part of the [event database](https://github.com/0chain/0chain/blob/88f91752f55d8957200fc07dab05cada98aa8b98/code/go/0chain.net/smartcontract/dbs/event/eventdb.go#L34),
```go
type EventDb struct {
	dbs.Store
	eventsChannel chan blockEvents
	settings      Settings
}
```
I could decouple this to a separate Pr, then these could be added to the `0chain.yaml` settings.

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 8]
```diff
@@ -577,7 +681,8 @@ func (edb *EventDb) addStat(event Event) error {
 		}
 		return edb.addOrUpdateChallengePools(*cps)
 	default:
-		return fmt.Errorf("unrecognised event %v", event)
+		logging.Logger.Debug("skipping event", zap.Int("tag", event.Tag))
+		return nil
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/dbs/event/process.go:715]: Is there a valid reason why we would be skipping an event here? 

This is a good spot to catch coding errors, so maybe we could recognise the events that can legitimately be skipped here and check this is one of them?

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 9]
```diff
@@ -296,6 +355,32 @@ func (edb *EventDb) processEvent(event Event, tags []int, round int64, block str
 	return tags, nil
 }
 
+func (edb *EventDb) updateSnapshots(e blockEvents, s *Snapshot) (*Snapshot, error) {
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/dbs/event/process.go:385]: Maybe this would fit better with the snapShot object as the receiver and the *EventDb as a parameter.

**Review by Sriep — COMMENTED** [review]
It's a big Pr, it might help to decouple any independent changes.

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feature/atlus_total_endpoints  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/snapshots  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go (hunk 4 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/blobber_aggregate.go]: yeah, good catch! we don't need this adjust for the last page anymore, since with 
```go
	size := int64(math.Ceil(float64(count) / float64(pageAmount)))
```
we guaranteed to have enough space to paginate count to exactly `pageAmount` pages

File: code/go/0chain.net/smartcontract/dbs/event/process.go (hunk 7 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/process.go:33]: we can change it in additional PR I believe after rewards are reviewed and merged

File: code/go/0chain.net/smartcontract/dbs/event/process.go (hunk 8 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/process.go:715]: we can have some special events that should be processed in Snapshots for example. In general we can use different type for them, but I don't see any reason we should process all the events here, in theory this is not the only possible processor, usually, they filter needed events and process them.

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/snapshot-batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/dbs/event/snapshot.go (hunk 5 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/snapshot.go:25]: @peterlimg my main concern about currency here that it's uint64 and sometimes aggregates can be negative, it's not coins, but some sort of derivative, so I left it as ints which are more like points, but not real tokens.

File: code/go/0chain.net/smartcontract/dbs/benchmark/setup.go [hunk 10]
```diff
@@ -56,8 +55,8 @@ func AddMockTransactions(
 	if !viper.GetBool(benchmark.EventDbEnabled) {
 		return
 	}
-	const txnTypeSmartContract = 1000
-	for blockNumber := int64(0); blockNumber <= viper.GetInt64(benchmark.NumBlocks); blockNumber++ {
+	const txnTypeStats = 1000
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/benchmark/setup.go]: Change back to txnTxnSmartContract to avoid confusion later

File: code/go/0chain.net/smartcontract/dbs/event/blobber.go [hunk 11]
```diff
@@ -48,6 +50,8 @@ type Blobber struct {
 	OffersTotal  currency.Coin `json:"offers_total"`
 	UnstakeTotal currency.Coin `json:"unstake_total"`
 	TotalStake   currency.Coin `json:"total_stake"`
+	//todo update
+	TotalServiceCharge currency.Coin `json:"total_service_charge"`
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/blobber.go:54]: When to update, will we do in a separate PR?

File: code/go/0chain.net/smartcontract/dbs/event/process.go [hunk 12]
```diff
@@ -374,15 +459,25 @@ func (edb *EventDb) addStat(event Event) error {
...
 		users, ok := fromEvent[[]User](event.Data)
 		if !ok {
 			return ErrInvalidEventData
 		}
-		return edb.addOrUpdateUsers(*users)
+		err := edb.addOrUpdateUsers(*users)
+		if err != nil {
+			for _, u := range *users {
+				b, _ := u.Balance.Int64()
+				logging.Logger.Debug("saving user", zap.String("id", u.UserID),
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/dbs/event/process.go]: This may flush the logs as a block can have lots of users.

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 13]
```diff
@@ -1093,6 +1127,10 @@ func (sc *StorageSmartContract) updateAllocationRequestInternal(
 		return "", common.NewError("allocation_updating_failed",
 			"can't get existing allocation: "+err.Error())
 	}
+	_, err = storageAllocationToAllocationTable(alloc)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: this function couldn't go wrong, so let's remove the returned error and checking

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 14]
```diff
@@ -1683,6 +1737,16 @@ func (sc *StorageSmartContract) finishAllocation(
 		return common.NewError("fini_alloc_failed",
 			"saving challenge pool: "+err.Error())
 	}
+	i, err := prevBal.Int64()
+	if err != nil {
+		return common.NewError("fini_alloc_failed",
+			"emitting pool change "+err.Error())
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: it's not an emitting error, let's use correct error message

File: code/go/0chain.net/smartcontract/storagesc/benchmark_setup.go [hunk 15]
```diff
@@ -422,9 +420,10 @@ func AddMockBlobbers(
 			}
 			blobberDb.TotalStake, err = currency.ParseZCN(viper.GetFloat64(sc.StorageMaxStake))
 			if err != nil {
-				panic(err)
+				log.Fatal("covert currency", err)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/benchmark_setup.go]: convert?

File: code/go/0chain.net/smartcontract/storagesc/stakepool.go [hunk 16]
```diff
@@ -150,6 +150,24 @@ func (sp *stakePool) stake() (stake currency.Coin, err error) {
...
+//	case spenum.Blobber:
+//		data.Updates["total_stake"] = int64(staked)
+//		balances.EmitEvent(event.TypeStats, event.TagUpdateBlobber, providerID, data)
+//	case spenum.Validator:
+//		data.Updates["stake"] = int64(staked)
+//		balances.EmitEvent(event.TypeStats, event.TagUpdateValidator, providerID, data)
+//	default:
+//		logging.Logger.Error("invalid providerType in stakepool StakeEvent")
+//	}
+//}
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/stakepool.go]: Remove if not used

File: code/go/0chain.net/smartcontract/storagesc/stakepool.go [hunk 17]
```diff
@@ -538,6 +563,8 @@ func (ssc *StorageSmartContract) stakePoolUnlock(
 		return "", common.NewErrorf("stake_pool_unlock_failed",
 			"stake pool staking error: %v", err)
 	}
+	//todo merge problem
+	//sp.emitSaveEvent(spr.ProviderType, spr.ProviderID, balances)
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/stakepool.go]: Should we create an issue to track this TODO?

**Review by peterlimg — CHANGES_REQUESTED** [review]
A few minor change requests

File: code/go/0chain.net/smartcontract/dbs/event/process.go (hunk 8 above)
Comment (Sriep) [reply code/go/0chain.net/smartcontract/dbs/event/process.go:715]: Yes, using the type to indicate which event processor to use seems like a good way to go. So if we get here, it means the event has the wrong type, or we forgot to write a handler for it.

**Comment by service-0chain** [comment]
Manual system tests [cancelled] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feature/snapshot-batch-update  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/snapshot-batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/dbs/event/process.go (hunk 8 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/process.go:715]: Hm, looks like it's stillnot good, imagine we have to different processor that have an intersection on event, should we have it in different types? not sure we should change it now

File: code/go/0chain.net/smartcontract/storagesc/stakepool.go (hunk 17 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/storagesc/stakepool.go]: no,it is updated inside `sp.stakeForProvider(...)`

File: code/go/0chain.net/smartcontract/dbs/event/blobber.go (hunk 11 above)
Comment (dabasov) [reply code/go/0chain.net/smartcontract/dbs/event/blobber.go:54]: yeah, most likely we need to sort it out in different issue. I'll try to fix it tomorrow, not sure I understand why we need total here, but not in BlobberAggregates. @Sriep might know

**Review by peterlimg — APPROVED** [review]
LGTM, please fix the system tests

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feature/atlus_total_endpoints  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/snapshot-batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | skip_api_blobber_rewards  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/snapshot-batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | feature/snapshot-batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |
